test/format-test.cc

```diff
@@ -2005,9 +2005,11 @@ class mock_arg_formatter
     : public fmt::internal::arg_formatter_base<buffer_range> {
  private:
 #if FMT_USE_INT128
-  MOCK_METHOD1(call, void(__int128_t value));
+  MOCK_METHOD1(callSigned, void(__int128_t value));
+  MOCK_METHOD1(callUnsigned, void(__uint128_t value));
 #else
-  MOCK_METHOD1(call, void(long long value));
+  MOCK_METHOD1(callSigned, void(long long value));
+  MOCK_METHOD1(callUnsigned, void(unsigned long long value));
```

Comment:
This is unnecessary for the test purposes. Just static cast to the largest signed type.

Reply:
The test is calling the mocked method with both signed and unsigned values.

I don't care as much about the tests being -Wsign-conversion clean, but...

I want my own code to be -Wsign-conversion clean, because we got bitten in the past by (unsigned)-1, and other bit fields confusion (somebody shoving 32 bits into a 16 bit register) and I want to use {fmt}. Which means {fmt} needs to be -Wsign-conversion clean, which means its test suite needs to be -Wsign-conversion clean, otherwise we can't enforce it via the automated regression suite.

Reply:
This is a simple mock which is only used here: https://github.com/fmtlib/fmt/blob/62da1db62affe82dc3ff5e5396e7b5cc235eca3d/test/format-test.cc#L2052

so casting is fine.

Reply:
I changed as suggested and I got back the dreaded:

fmt/test/format-test.cc:2026:10: error: implicit conversion changes signedness: 'unsigned __int128' to '::testing::internal::Function<void (__int128_t)>::Argument1' (aka '__int128') [-Werror,-Wsign-conversion]

Reply:
This is the long story, threaded through format.h and core.h:

> /home/florin/libraries/fmt/test/format-test.cc: In instantiation of ‘typename std::enable_if<fmt::v6::internal::is_integral<T>::value, std::back_insert_iterator<fmt::v6::internal::buffer<char> > >::type mock_arg_formatter::operator()(T) [with T = __int128 unsigned; typename std::enable_if<fmt::v6::internal::is_integral<T>::value, std::back_insert_iterator<fmt::v6::internal::buffer<char> > >::type = std::back_insert_iterator<fmt::v6::internal::buffer<char> >]’:

> /home/florin/libraries/fmt/include/fmt/core.h:1024:15:   required from ‘constexpr decltype (vis(0)) fmt::v6::visit_format_arg(Visitor&&, const fmt::v6::basic_format_arg<Context>&) [with Visitor = mock_arg_formatter; Context = fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::internal::buffer<char> >, char>; decltype (vis(0)) = std::back_insert_iterator<fmt::v6::internal::buffer<char> >]’

> /home/florin/libraries/fmt/include/fmt/format.h:3172:25:   required from ‘void fmt::v6::format_handler<ArgFormatter, Char, Context>::on_replacement_field(const Char*) [with ArgFormatter = mock_arg_formatter; Char = char; Context = fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::internal::buffer<char> >, char>]’

> /home/florin/libraries/fmt/include/fmt/format.h:2583:7:   required from ‘constexpr void fmt::v6::internal::parse_format_string(fmt::v6::basic_string_view<Char>, Handler&&) [with bool IS_CONSTEXPR = false; Char = char; Handler = fmt::v6::format_handler<mock_arg_formatter, char, fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::internal::buffer<char> >, char> >&]’

> /home/florin/libraries/fmt/include/fmt/format.h:3205:39:   required from ‘typename Context::iterator fmt::v6::vformat_to(typename ArgFormatter::range, fmt::v6::basic_string_view<Char>, fmt::v6::basic_format_args<Context>, fmt::v6::internal::locale_ref) [with ArgFormatter = mock_arg_formatter; Char = char; Context = fmt::v6::basic_format_context<std::back_insert_iterator<fmt::v6::internal::buffer<char> >, char>; typename Context::iterator = std::back_insert_iterator<fmt::v6::internal::buffer<char> >; typename ArgFormatter::range = fmt::v6::buffer_range<char>]’

> /home/florin/libraries/fmt/test/format-test.cc:2043:63:   required from here
> /home/florin/libraries/fmt/test/format-test.cc:2026:10: error: conversion to ‘testing::internal::Function<void(__int128)>::Argument1’ {aka ‘__int128’} from ‘__int128 unsigned’ may change the sign of the result [-Werror=sign-conversion]
 2026 |     call(value);
      |          ^~~~~